rancher2/resource_rancher2_cluster.go

```diff
@@ -649,72 +649,75 @@ func isKubeConfigValid(c *Config, config string) (string, bool, error) {
 func isKubeConfigTokenValid(c *Config, config string) (string, bool, error) {
 	token, err := getTokenFromKubeConfig(config)
 	if err != nil {
-		return "", false, fmt.Errorf("Getting Kubeconfig token: %v", err)
+		return "", false, fmt.Errorf("getting Kubeconfig token: %v", err)
 	}
 	isValid, err := isTokenValid(c, splitTokenID(token))
 	if err != nil {
-		return "", false, fmt.Errorf("Checking Kubeconfig token: %v", err)
+		return "", false, fmt.Errorf("checking Kubeconfig token: %v", err)
 	}
 	return token, isValid, nil
 }
 
-func replaceKubeConfigToken(c *Config, config, token string) (string, error) {
-	if len(token) == 0 {
-		return config, nil
-	}
+func replaceKubeConfigToken(c *Config, config string) (string, error) {
 	kubeconfig, err := getObjFromKubeConfig(config)
 	if err != nil {
-		return "", fmt.Errorf("Getting K8s config object: %v", err)
+		return "", fmt.Errorf("getting K8s config object: %v", err)
 	}
 	if kubeconfig == nil || kubeconfig.AuthInfos == nil || len(kubeconfig.AuthInfos) == 0 {
 		return config, nil
 	}
 
 	client, err := c.ManagementClient()
 	if err != nil {
-		return "", fmt.Errorf("Replacing cluster Kubeconfig token: %v", err)
+		return "", fmt.Errorf("replacing cluster Kubeconfig token: %v", err)
 	}
-	removeToken, err := client.Token.ByID(splitTokenID(kubeconfig.AuthInfos[0].AuthInfo.Token))
-	if err != nil {
+
+	// if cached token is corrupt, ByID will fail and the token can't be
+	// retrieved or cleaned up
+	currentToken, err := client.Token.ByID(splitTokenID(kubeconfig.AuthInfos[0].AuthInfo.Token))
+	if err != nil || currentToken.Expired {
 		if !IsNotFound(err) && !IsForbidden(err) {
 			return "", err
 		}
+		// client can't find the token or token has expired, so create
+		// a new one
+		currentToken, err = client.Token.Create(currentToken)
+		if err != nil {
+			return "", fmt.Errorf("error creating Token: %s", err)
+		}
 	}
-
-	err = client.Token.Delete(removeToken)
```

Comment:
Not a review comment but question, we've confirmed that expired tokens get cleaned up by rancher and we don't need to delete it here right?